azurerm/internal/services/devtestlabs/resource_arm_dev_test_global_shutdown_schedule.go

```diff
@@ -0,0 +1,303 @@
+package devtestlabs
+
+import (
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/devtestlabs/mgmt/2016-05-15/dtl"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/features"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/devtestlabs/parse"
+	devtestValidate "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/devtestlabs/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tags"
+	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmDevTestLabGlobalShutdownSchedule() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmDevTestLabGlobalShutdownScheduleCreateUpdate,
+		Read:   resourceArmDevTestLabGlobalShutdownScheduleRead,
+		Update: resourceArmDevTestLabGlobalShutdownScheduleCreateUpdate,
+		Delete: resourceArmDevTestLabGlobalShutdownScheduleDelete,
+		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
+			_, err := parse.GlobalScheduleID(id)
+			return err
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"location": azure.SchemaLocation(),
+
+			"virtual_machine_id": {
```

Comment:
it seems like this resource acts on a vm? not globally as the name implies?

should this resource instead be called `azurerm_dev_test_vm_global_shutdown_schedule` to highlight it works on a vm?

Reply:
The global part comes from the underlying API. I think it's named that because it actually operates outside the context of DevTestLabs, but the API is still under the Microsoft.DevTestLabs provider. Thus it's a "global" schedule in the sense that it's not local to DevTestLabs.

I'm really not a fan of the name as is honestly. In my opinion, ideally the resource name wouldn't even have `dev_test` in it because it's not really applicable to DevTestLabs, it just uses the same resource provider. In an ideal world, I would name it `azurerm_virtual_machine_shutdown_schedule` so that it would be more easily discoverable and intuitive to find. I know that if I was looking through the documentation and saw a resource `azurerm_dev_test_vm_global_shutdown_schedule` I would not think that it would apply to normal Azure compute VMs, but it does.

I don't know the project's appetite for breaking the namespacing of resources like that. If that's too much, I'm fine with your proposed resource name. I might just propose switching the order to `global_vm_shutdown_schedule` to make "global" parse as a modifier for "vm" to better indicate it's not referencing normal DevTestLabs VMs.

I will work on the other changes you've requested to clean up the code a bit. Once that's done I can update the name however you decide.

Reply:
@sean-nixon - entirely agree, the API and SDK naming can be pretty awful. We break it all the time for what will make the most sense, and typically try to follow the portal over the API. i think `azurerm_dev_test_global_vm_shutdown_schedule` is a good compromise, as it is part of the dev test labs service.